### PR TITLE
Decouple Cell and CrossSection

### PR DESCRIPTION
**Pull request by flaport**
This PR decouples the CrossSection from the Cell. This has the advantage that a CrossSection will no longer have a z_min and z_max (present in Cell) which should improve caching of previously calculated modes.

This required some breaking changes:

1. Cell is no longer an attribute of CrossSection. Wherever you used `cs.cell` or similar you'll have to find an alternative way of accessing the needed information. Note: `mesh` is now a native attribute of `CrossSection`.
2. Use `CrossSection.from_cell(cell=cell, env=env)` wherever you used to use `CrossSection(cell=cell, env=env)`
3. A `Mode` depends on a `CrossSection` not on a `Cell` and hence its `.cell` attribute is now also gone. This will obviously improve caching but unfortunately this also means you'll have to provide `cells` as a second argument to `compute_s_matrix` from now on. Since `compute_s_matrix` only depends on the cell length you can also skip supplying `cells` and specify `cell_lengths` as a keyword argument for that function.
4. The `ez_interfaces` attribute (previously in `Cell`) is now moded to `Mesh`.

There are a bunch of other minor changes under the hood (such as Geometry->Geometry2D,Geometry3D and Structure->Structure2D,Structure3D) but these should not touch the user API.

@Jan-David-Black , feel free to comment or request changes. I'm planning to merge this branch on Thursday.

### CODE REVIEW

**Comment by jan-david-fischbach** [comment]
@flaport sounds good to me. I'll have to get my hands dirty with it though to say more :) Hopefully I'll find time before Thursday!

File: meow/cell.py [hunk 1]
```diff
@@ -1,39 +1,32 @@
...
+    Structure2D,
+    Structure3D,
     classify_structures_by_mesh_order_and_material,
     sort_structures,
 )
 
 
 class Cell(BaseModel):
-    """A `Cell` defines a location in a `Structure` associated with a `Mesh`"""
+    """A `Cell` defines a location in a `Structure3D` associated with a `Mesh`"""
```
Comment (jan-david-fischbach) [inline meow/cell.py]: Maybe something like
> A `Cell` defines an interval in z (the direction of propagation) within the simulation domain. The intersecting `Structure3D`s are discretized by a given mesh at the center of the `Cell`

File: meow/cell.py [hunk 2]
```diff
@@ -132,21 +124,43 @@ def create_cells(
...
 
-def _create_material_array(cell, structures, ez_interfaces):
-    m_full = np.zeros_like(cell.mesh.X_full, dtype=np.int_)
+def _create_full_material_array(
+    mesh: Mesh2D,
+    structures: List[Structure2D],
+    materials: Dict[Material, int],
+):
+    m_full = np.zeros_like(mesh.X_full, dtype=np.int_)
+    structures_dict = classify_structures_by_mesh_order_and_material(
```
Comment (jan-david-fischbach) [inline meow/cell.py:142]: Reminder to self: potential problem with dict ordering

File: meow/eme/propagate.py [hunk 3]
```diff
@@ -103,18 +102,17 @@ def compute_mode_amplitudes(u, v, m, excitation_l, excitation_r):
     return forward, backward
 
 
-def plot_fields(modes, forwards, backwards, y, z, lim=1):
+def plot_fields(modes, cells, forwards, backwards, y, z, lim=1):
```
Comment (jan-david-fischbach) [inline meow/eme/propagate.py:105]: Sidenote independent of PR. Should make this compatible to multiple meshes along z?

File: meow/structures.py [hunk 4]
```diff
@@ -41,13 +90,44 @@ def visualize_structures(structures, scale=None):
...
 
-def classify_structures_by_mesh_order_and_material(structures, materials):
+def classify_structures_by_mesh_order_and_material(
+    structures: Union[List[Structure3D], List[Structure2D]],
+    materials: Dict[Material, int],
+) -> Union[
+    Dict[Tuple[int, int], List[Structure2D]], Dict[Tuple[int, int], List[Structure3D]]
+]:
     structures = sort_structures(structures)
     structures_dict = {}
```
Comment (jan-david-fischbach) [inline meow/structures.py:132]: The order of this dict determines the layering of structures while building the material matrix. It should this be stored in a datastructure, that respects order (e.g. `OrderedDict`)

**Review by jan-david-fischbach — COMMENTED** [review]
My first ideas

File: meow/cell.py (hunk 2 above)
Comment (flaport) [reply meow/cell.py:142]: python dicts are guarenteed to be ordered since python 3.6

File: meow/structures.py (hunk 4 above)
Comment (flaport) [reply meow/structures.py:132]: a python dictionary is guarenteed to be ordered since python 3.6

File: meow/eme/propagate.py (hunk 3 above)
Comment (flaport) [reply meow/eme/propagate.py:105]: not sure what you mean by that?

File: meow/cell.py (hunk 2 above)
Comment (jan-david-fischbach) [reply meow/cell.py:142]: Seems I am not up to date then... Sorry

File: meow/structures.py (hunk 4 above)
Comment (jan-david-fischbach) [reply meow/structures.py:132]: Same here

File: meow/eme/propagate.py (hunk 3 above)
Comment (jan-david-fischbach) [reply meow/eme/propagate.py:105]: I get the mesh from the first cell and assume that the mesh is the same for all cells. However, I have seen that the option exists to have a different mesh on every Cell. (How do we calculate the overlap in these cases btw?)

Comment (flaport) [reply meow/eme/propagate.py:105]: Yeah, I left the option open for different meshes in each cell, but it's clearly not fully supported. I think to calculate the overlap you probably have to interpolate the modes from the smaller mesh to fit on the bigger mesh or something like that. In any case... not really a priority right now.

That said, maybe a simple check to ensure all the meshes are the same would be good?

Comment (jan-david-fischbach) [reply meow/eme/propagate.py:105]: Jep, seems like a good check to avoid unwanted behavior. I think we should also check in the overlap calculations

**Comment by jan-david-fischbach** [comment]
👍 Do we delete the branch?

**Pull request merged**